include/onnxruntime/core/framework/data_types.h

```diff
@@ -358,6 +381,36 @@ struct SetSequenceType {
   }
 };
 
+/// Optional helpers
+
+void CopyMutableOptionalElement(const ONNX_NAMESPACE::TypeProto&,
+                                ONNX_NAMESPACE::TypeProto&);
+
+template <typename T, typename elemT>
+struct SetOptionalType {
+  static void Set(ONNX_NAMESPACE::TypeProto& proto) {
+    const onnx::TypeProto* elem_proto = nullptr;
+    if (std::is_same<T, Tensor>::value) {
+      MLDataType dt = DataTypeImpl::GetTensorType<elemT>();
+      elem_proto = dt->GetTypeProto();
+    } else if (std::is_same<T, TensorSeq>::value) {
+      MLDataType dt = DataTypeImpl::GetSequenceTensorType<elemT>();
+      elem_proto = dt->GetTypeProto();
+    } else {
+      // Will not reach here
+      ORT_ENFORCE(false, "Unsupproted type for optional type");
+    }
+
+#ifdef ORT_NO_RTTI
+    ORT_ENFORCE(elem_proto != nullptr, "expected a registered ORT type");
+#else
+    ORT_ENFORCE(elem_proto != nullptr, typeid(T).name(),
+                " expected to be a registered ORT type");
+#endif
+    CopyMutableOptionalElement(*elem_proto, proto);
+  }
+};
```

Comment:
can we restructure it as functor struct template specializations so we can move implementation into .cc file?

Reply:
I prefer not to do such refactoring in this PR. It is going to make reviewing this change a bit more harder. We can always refactor post this PR.